File: 921-spiral-matrix-iii/spiral-matrix-iii.cpp
class Solution {
public:
    vector<vector<int>> spiralMatrixIII(int rows, int cols, int rStart, int cStart) {
        vector<vector<int>> dir = {{0,1}, {1,0}, {0,-1}, {-1,0}};
        vector<vector<int>> res;
        int steps = 1;
        int r = rStart, c = cStart;
        int d = 0; // direction index

        while (res.size() < rows * cols) {
            for (int j = 0; j < 2; ++j) {
                for (int i = 0; i < steps; ++i) {
                    if (0 <= r && r < rows && 0 <= c && c < cols) {
                        res.push_back({r, c});
                    }
                    r += dir[d][0];
                    c += dir[d][1];
                }
                d = (d + 1) % 4; 
            }
            steps++;
        }
        return res;
    }
};

File: 921-spiral-matrix-iii/spiral-matrix-iii.py
class Solution:
    def spiralMatrixIII(self, rows: int, cols: int, rStart: int, cStart: int) -> List[List[int]]:
        dir = [[0,1],[1,0],[0,-1],[-1,0]]
        res = []
        step = 1
        r,c=rStart, cStart
        i=0     # to keep track of direction
        while(len(res) < rows*cols):
            for _ in range(2):
                dr,dc = dir[i]
                for y in range(step):
                    if(0<=r<rows and 0<=c<cols):
                        res.append([r,c])
                    r,c = r+dr, c+dc

                i = (i+1)%4
            step+=1

        return res